pymeasure/adapters/adapter.py

```diff
@@ -33,48 +37,135 @@ class Adapter:
 
     This class should only be inherited from.
 
-    :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+    :param preprocess_reply: (deprecated) optional callable used to preprocess
+        strings received from the instrument. The callable returns the
+        processed string.
+    :param float query_delay: Time in s to wait after writing and before reading.
+    :param log: Parent logger of the 'Adapter' logger.
     :param kwargs: all other keyword arguments are ignored.
     """
 
-    def __init__(self, preprocess_reply=None, **kwargs):
+    def __init__(self, preprocess_reply=None, query_delay=0, log=None, **kwargs):
         self.preprocess_reply = preprocess_reply
+        self.query_delay = query_delay
         self.connection = None
+        if log is None:
+            self.log = logging.getLogger("Adapter")
+        else:
+            self.log = log.getChild("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn("Deprecated in Adapter, pass it to Instrument.__init__, instead.",
+                 FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
+        self.close()
+
+    def close(self):
+        """Close the connection."""
         if self.connection is not None:
             self.connection.close()
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
+    def write(self, command, **kwargs):
+        """Write a string command to the instrument appending `write_termination`.
 
-        :param command: SCPI command string to be sent to the instrument
+        Do not override in a subclass!
+
+        :param str command: Command string to be sent to the instrument
+            (without termination).
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", command))
+        self._write(command, **kwargs)
+
+    def write_bytes(self, content, **kwargs):
+        """Write the bytes `content` to the instrument.
+
+        Do not override in a subclass!
+
+        :param bytes content: The bytes to write to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", content))
+        self._write_bytes(content, **kwargs)
+
+    def wait_till_read(self, query_delay=0):
```

Comment:
I am not sure whether this method is useful, I see it as a feature of the instrument and not of the adapter.

Reply:
The idea was, that it might be a feature of the instrument and / or of the adapter, for even an adapter might require a delay. 
In order to enable adapter delays, either the instrument reads the adapter value, or (implemented here) the instrument tells the adapter its own delay.